### PR TITLE
Reference implementations for rot90, roll, atleast_1d,2d,3d

### PR DESCRIPTION
**Pull request by IvanYashchuk**
This PR adds the following references:

- `rot90`
- `roll`
- `atleast_1d`
- `atleast_2d`
- `atleast_3d`

### CODE REVIEW

File: torch/_refs/__init__.py [hunk 1]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
     return _reshape_view_helper(a, shape, allow_copy=True)
 
 
+def rot90(a: TensorLikeType, k: int = 1, dims: DimsSequenceType = (0, 1)) -> TensorLikeType:
+    """Reference implementation of :func:`torch.rot90`."""
+    if len(dims) != 2:
+        raise ValueError(f"expected total rotation dims == 2, but got dims = {len(dims)}")
+    if a.ndim < 2:
+        raise ValueError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and (dims[0] - dims[1]).__abs__() != a.ndim):
```
Comment (mruberry) [inline torch/_refs/__init__.py:1588]: Regular Python abs is probably more readable than the dunder here?

File: torch/_refs/__init__.py [hunk 2]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
     return _reshape_view_helper(a, shape, allow_copy=True)
 
 
+def rot90(a: TensorLikeType, k: int = 1, dims: DimsSequenceType = (0, 1)) -> TensorLikeType:
+    """Reference implementation of :func:`torch.rot90`."""
+    if len(dims) != 2:
```
Comment (mruberry) [inline torch/_refs/__init__.py:1584]: I think this needs to canonicalize the dims upfront with utils.canonicalize_dims (which wraps dims specified as negative numbers)

File: torch/_refs/__init__.py [hunk 3]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+        raise ValueError(f"expected total rotation dims == 2, but got dims = {len(dims)}")
+    if a.ndim < 2:
+        raise ValueError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and (dims[0] - dims[1]).__abs__() != a.ndim):
+        raise ValueError(f"expected rotation dims to be different, but got dim0 = {dims[0]} and dim1 = {dims[1]}")
+    if not (dims[0] < a.ndim and dims[0] >= -a.ndim):
```
Comment (mruberry) [inline torch/_refs/__init__.py:1590]: dim canonicalization should take care of these checks, too

File: torch/_refs/__init__.py [hunk 4]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+
+        if len_shifts != len_dims:
+            raise ValueError(f"shifts and dimensions must align. shifts: {len_shifts}, dims: {len_dims}")
+        assert len_dims > 1
+        tail_shifts = shifts[1:]
+        tail_dims = dims[1:]
+        first_dim_rolled = roll(a, shifts[0], dims[0])
+        return roll(first_dim_rolled, tail_shifts, tail_dims)
+
+    if a.numel == 0:
```
Comment (mruberry) [inline torch/_refs/__init__.py:1629]: Comment for this sho-circuit -- should this be sooner?

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+        return clone(a)
+
+    dim = dims[0]
+    size = a.shape[dim]
+    start = (size - shifts[0]) % size
+    t0 = narrow(a, dim, start, size - start)
+    t1 = narrow(a, dim, 0, start)
+    return cat((t0, t1), dim)
```
Comment (mruberry) [inline torch/_refs/__init__.py:1637]: Wow this is a crazy function

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -6718,11 +6718,9 @@ def sample_inputs_roll(op_info, device, dtype, requires_grad=False, **kwargs):
 def sample_inputs_rot90(op_info, device, dtype, requires_grad=False, **kwargs):
     make_arg = partial(make_tensor, device=device, dtype=dtype, requires_grad=requires_grad)
 
-    args = ((1, (0, 1),),
-            (1, (1, 2),),
-            (1, (1, -1),),
-            ())
+    args = itertools.product(range(-5, 6), [(0, 1), (1, 2), (1, -1)])
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:6721]: Nice

File: torch/testing/_internal/common_methods_invocations.py [hunk 7]
```diff
@@ -19230,6 +19240,14 @@ def __init__(
         "_refs.reshape",
         torch_opinfo_name="reshape",
     ),
+    PythonRefInfo(
+        "_refs.rot90",
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:19244]: This looks really good -- are the sample inputs for rot90 and/or roll sufficient, or should we use this opportunity to extend them with some reference inputs, too, to ensure all code paths and edge cases are validated? 

Also, although it is more work, error inputs for rot90 and roll would be great to verify their many error checks, but your call whether it's worth going there at this time

**Comment by mruberry** [comment]
Overall looks good, @IvanYashchuk -- just a few questions and suggestions inline, and we should wait for the tests. Looking forward to hearing your thoughts on adding references!

File: torch/_refs/__init__.py (hunk 1 above)
Comment (mruberry) [reply torch/_refs/__init__.py:1588]: I know and that's lame but we should get consistent about calling `builtins.abs` directly (and more generally calling `builtins.foo` whenever we clobber a Python name)

Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1588]: Thanks! Let's use `builtins.abs`.

File: torch/_refs/__init__.py (hunk 2 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1584]: Initially I thought that `canonicalize_dims` is not needed here because `dims` is passed to `flip` and `transpose` which themselves call `canonicalize_dims` taking care of negative numbers. But then I see now that the "out of range" checks are unnecessarily duplicated in this function.

Comment (mruberry) [reply torch/_refs/__init__.py:1584]: Eventually we should investigate an error context so even if an error is thrown from another operation it cites the user-facing entrypoint as expected.

However, we don't have an error context today, so I would suggest duplicating error checks for the moment both for message quality and for code simplicity -- because it can sometimes be tricky to reason about what other operations will or will not validate

**Comment by IvanYashchuk** [comment]
I added error inputs for `roll` and `rot90` and addressed review feedback. It seems the XLA failure is unrelated.

@mruberry, could you please take another look?

File: torch/_refs/__init__.py [hunk 8]
```diff
@@ -1552,6 +1579,64 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+    if not isinstance(dims, Iterable):
+        dims = (dims,)
+
+    len_shifts = len(shifts)
+    len_dims = len(dims)
+    if len_shifts != 1 or len_dims != 1:
+        if len_shifts == 0:
+            raise ValueError("`shifts` required")
+        if len_dims == 0 and len_shifts == 1:
+            return clone(view(roll(flatten(a), shifts, 0), a.shape))
```
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/_refs/__init__.py:1857]: If clone is just to pacify the tests, maybe we should disable them (we disable them for reduction ops e.g.)

File: torch/_refs/__init__.py [hunk 9]
```diff
@@ -1784,6 +1830,80 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+    # Required to silence MyPy errors
+    assert isinstance(dims_, (tuple, list))
+    dims = dims_
+    if len(dims) != 2:
+        raise RuntimeError(
+            f"expected total rotation dims == 2, but got dims = {len(dims)}"
+        )
+    if a.ndim < 2:
+        raise RuntimeError(f"expected total dims >= 2, but got total dims = {a.ndim}")
+    if not (dims[0] != dims[1] and builtins.abs(dims[0] - dims[1]) != a.ndim):
```
Comment (ngimel) [inline torch/_refs/__init__.py:1892]: second part of conditional cannot happen after canonicalization?

File: torch/_refs/__init__.py [hunk 10]
```diff
@@ -1784,6 +1830,80 @@ def reshape(a: TensorLikeType, shape: ShapeType) -> TensorLikeType:
...
+        raise RuntimeError(
+            f"expected rotation dims to be different, but got dim0 = {dims[0]} and dim1 = {dims[1]}"
+        )
+    k = k % 4  # Rotation direction is from the second towards the first axis for k < 0
+    if k == 1:
+        return clone(transpose(flip(a, (dims[1],)), dims[0], dims[1]))
```
Comment (ngimel) [inline torch/_refs/__init__.py:1898]: are clones here again to pacify view tests?

File: torch/_refs/__init__.py (hunk 9 above)
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1892]: Right. I will remove it.

**Review by mruberry — APPROVED** [review]
Test failures are unrelated -- I think this is good to go!

Thanks @IvanYashchuk -- curious to hear your thoughts about where we should go next

**Comment by mruberry** [comment]
@ezyang are we auto-registering something internally that could have caused this to break XLA? I don't think we're running any primTorch tests on XLA

**Comment by JackCaoG** [comment]
Thanks, I just saw the XLA test failed at `BuildRoll`. Feel free to open a ticket on xla side, I will take a look why test crashed.

**Comment by IvanYashchuk** [comment]
@pytorchbot merge this

backwards_compat failure is unrelated:
```
Broken ops: [
	aten::special_laguerre_polynomial_l(Tensor x, Tensor n) -> Tensor
	aten::special_laguerre_polynomial_l.out(Tensor x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.n_scalar_out(Tensor x, Scalar n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.x_scalar(Scalar x, Tensor n) -> Tensor
	aten::special_laguerre_polynomial_l.x_scalar_out(Scalar x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_laguerre_polynomial_l.n_scalar(Tensor x, Scalar n) -> Tensor
	aten::special_hermite_polynomial_he(Tensor x, Tensor n) -> Tensor
	aten::special_hermite_polynomial_he.out(Tensor x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.n_scalar_out(Tensor x, Scalar n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.x_scalar(Scalar x, Tensor n) -> Tensor
	aten::special_hermite_polynomial_he.x_scalar_out(Scalar x, Tensor n, *, Tensor(a!) out) -> Tensor(a!)
	aten::special_hermite_polynomial_he.n_scalar(Tensor x, Scalar n) -> Tensor
	aten::_linalg_svd(Tensor A, bool full_matrices=False, bool compute_uv=True, *, str? driver=None) -> (Tensor U, Tensor S, Tensor Vh)
	aten::_linalg_svd.U(Tensor A, bool full_matrices=False, bool compute_uv=True, *, str? driver=None, Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh) -> (Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh)
	aten::linalg_svd(Tensor A, bool full_matrices=True, *, str? driver=None) -> (Tensor U, Tensor S, Tensor Vh)
	aten::linalg_svd.U(Tensor A, bool full_matrices=True, *, str? driver=None, Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh) -> (Tensor(a!) U, Tensor(b!) S, Tensor(c!) Vh)
	aten::linalg_svdvals(Tensor A, *, str? driver=None) -> Tensor
	aten::linalg_svdvals.out(Tensor A, *, str? driver=None, Tensor(a!) out) -> Tensor(a!)
]
```
